tests/optim/test_oss.py

```diff
@@ -334,3 +334,62 @@ def test_collect_shards():
     mp.spawn(
         run_test_collect_shards, args=(world_size, reference_rank), nprocs=world_size, join=True,
     )
+
+
+def run_test_multiple_groups(rank, world_size):
+    # Only work with the even ranks, to check that the global_rank indexing is properly used
+    dist_init(rank, world_size)
+    sub_group_ranks = [0, 2, 4]
+    pg = torch.distributed.new_group(ranks=sub_group_ranks, backend="gloo")
+    device = "cpu"
+    epochs, batch, input_width, hidden, target_width = 5, 3, 20, 10, 5
+    loss_fn = torch.nn.L1Loss().to(device)
+
+    def check(optimizer):
+        # Just run a couple of random epochs, check that the model is properly updated
+        for _ in range(epochs):
+            target = torch.rand((batch, target_width), device=device)
+            inputs = torch.rand((batch, input_width), device=device)
+
+            def closure():
```

Comment:
The test might be more realistic if you used DDP to reduce gradients.

Reply:
I don't get that, could you elaborate ?

Reply:
one issue with DDP is that I think that it requires CUDA, so it's more complicated to test with 3+ gpus so that there's a big enough sub-group where the optimizer works. Another point is that this is just meant to test the fact that multiple groups are properly handled by OSS, it's not a performance test, so I don't understand what DDP brings.

Reply:
I think it should be possible to all_reduce CPU tensors and should only add 1  or 2 more lines of code.

Tests also serve as examples of how to use the API so its nice if they match how the API is used in practice (when possible). If it only requires 1 or 2 more LOC it might be worth it.

Reply:
just to be sure, you mean all_reducing on the loss so that it looks like a normal training loop, or reducing all the gradients from here ?

Reply:
Agree with @msbaines. The challenge was getting this to work for DDP+MP, so the test should check for these combined.
@msbaines can also comment but I think all reducing on the loss is most realistic

Reply:
The committed version reduces on the loss only (and does catch your problem), we sync'ed 1:1 with Mandeep but that was too late, I initially missed it. I think that I understand that DDP would be nice here also, it needs to be there in the benchmark and tutorials in any case, but in this test it restricts the unit testing to platforms with 3+ GPUs, which are not that common. 

I can write a follow up unit test with DDP which only triggers if the HW fits, so that at least the CI catches it